[SPARK-32588][CORE][TEST] Fix SizeEstimator initialization in tests

mundaym opened the pull request:
In order to produce consistent results from SizeEstimator the tests
override some system properties that are used during SizeEstimator
initialization. However there were several places where either the
compressed references property wasn't set or the system properties
were set but the SizeEstimator not re-initialized.

This caused failures when running the tests with a large heap build
of OpenJ9 because it does not use compressed references unlike most
environments.

### What changes were proposed in this pull request?
Initialize SizeEstimator class explicitly in the tests where required to avoid relying on a particular environment.

### Why are the changes needed?
Test failures can be seen when compressed references are disabled (e.g. using an OpenJ9 large heap build or Hotspot with a large heap).

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Tests run on machine running OpenJ9 large heap build.

mundaym committed:
[SPARK-32588][TEST] Fix SizeEstimator initialization in tests.

In order to produce consistent results from SizeEstimator the tests
override some system properties that are used during SizeEstimator
initialization. However there were several places where either the
compressed references property wasn't set or the system properties
were set but the SizeEstimator not re-initialized.

This caused failures when running the tests with a large heap build
of OpenJ9 because it does not use compressed references unlike most
environments.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
Jenkins test this please

SparkQA commented:
**[Test build #127327 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127327/testReport)** for PR 29407 at commit [`1715d66`](https://github.com/apache/spark/commit/1715d66ec4c2c3355050dcbd92a876e3eb5339c9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31946/
Test PASSed.

SparkQA commented:
**[Test build #127327 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127327/testReport)** for PR 29407 at commit [`1715d66`](https://github.com/apache/spark/commit/1715d66ec4c2c3355050dcbd92a876e3eb5339c9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127327/
Test PASSed.

maropu commented:
cc: @kiszk @rednaxelafx

rednaxelafx commented on core/src/test/scala/org/apache/spark/util/SizeEstimatorSuite.scala:
```diff
@@ -214,6 +216,10 @@ class SizeEstimatorSuite
   }
 
   test("class field blocks rounding on 64-bit VM without useCompressedOops") {
+    System.setProperty(TEST_USE_COMPRESSED_OOPS_KEY, "false")
```
Since this PR is all about explicit re-initialization, should we add `System.setProperty("os.arch", "amd64")` to this block as well?
Actually would it be a good idea to wrap this sequence into a helper method or something?

kiszk replied:
I think that `System.setProperty("os.arch", "amd64")` has already done at `beforeEach()`.

rednaxelafx replied:
In which case it'd be better to keep all the relevant test cases consistent by removing those explicit initialization lines in a couple of other test cases in the same file?

mundaym replied:
I think we should probably be as explicit as possible about what each test is doing. Let me add a follow up commit using a helper function and you can see what you think.

rednaxelafx reviewed (commented):
Mostly LGTM. Thanks for making the tests more reliable!

One related question _outside_ the scope of this PR: since some system properties affect the behavior of the `SizeEstimator`, should the test suites that re-initialize `SizeEstimator` be responsible to restoring those properties back to the default (whatever state that came in before the suite) so that other suites are less likely to be affected?

mundaym commented:
> One related question outside the scope of this PR: since some system properties affect the behavior of the SizeEstimator, should the test suites that re-initialize SizeEstimator be responsible to restoring those properties back to the default (whatever state that came in before the suite) so that other suites are less likely to be affected?

I did wonder that myself. I don't know enough about how much of the environment is reset (new JVM instance?) between runs of each test suite to say either way really. Most probably the individual tests should reset everything to avoid accidental dependencies between them (e.g. a situation where one test sets a property and another test accidentally relies on it being set).

mundaym committed:
[SPARK-32588][FOLLOWUP][TEST] Refactor SizeEstimator initialization

Consolidate SizeEstimator initialization in SizeEstimatorSuite by
using a helper function as suggested by feedback in PR #29407.

srowen commented:
It's best to un-set whatever you set in tests, if it's global like a system property. The JVM is shared.

srowen commented:
Do we need to unset these properties in afterEach, per last comments? that is tidier.

mundaym committed:
[SPARK-32588][FOLLOWUP][TEST] Restore system properties and SizeEstimator after tests

System properties and the SizeEstimator class are modified by tests
in these suites. These global variables should be restored to their
original values after each test in order to prevent the modifications
affecting other tests.

mundaym commented:
> Do we need to unset these properties in afterEach, per last comments? that is tidier.

Added a new commit which does this. Please take a look when you get a chance. Thanks!

srowen commented on core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala:
```diff
@@ -144,10 +144,28 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with BeforeAndAfterE
+  val originalCompressedOops = System.getProperty(TEST_USE_COMPRESSED_OOPS_KEY)
+
+  def reinitializeSizeEstimator(arch: String, useCompressedOops: String): Unit = {
+    val set = (k: String, v: String) => {
```
Minor but, just a `def`?

mundaym replied:
Will do (not really a Scala developer!).

srowen commented on core/src/test/scala/org/apache/spark/storage/MemoryStoreSuite.scala:
```diff
@@ -51,12 +52,34 @@ class MemoryStoreSuite
-    System.setProperty("os.arch", "amd64")
-    val initialize = PrivateMethod[Unit](Symbol("initialize"))
-    SizeEstimator invokePrivate initialize()
+    reinitializeSizeEstimator("amd64", "true")
```
Should this second arg be true here? Some tests that don't set this currently now set it to true, some to false.

mundaym replied:
Yeah, this test suite assumes that compressed oops are enabled. The comment above alludes to this but the original code didn't actually override it. This leads to test failures when running a JVM where compressed oops are disabled.

In another test ("64-bit arch with no compressed oops") the test was inheriting the compressed oops setting of false from a preceding test rather than setting it itself. Since the preceding test is now cleaning up after itself that test now explicitly needs to set compressed oops to false.

In general it seems better to explicitly set these settings. Or we could try and make the tests more robust to different arch and compressed oops combinations. That is a bit more involved than this patch though.

srowen replied:
OK if they're all on purpose and as intended, that's fine.

kiszk commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32320/
Test PASSed.

SparkQA commented:
**[Test build #127696 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127696/testReport)** for PR 29407 at commit [`a5c158e`](https://github.com/apache/spark/commit/a5c158e9dfeb0fc9e5a57f172a5780f8cae6a17b).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127696/
Test FAILed.

srowen commented:
Looks OK @kiszk ?

kiszk commented:
Let me see later.

srowen commented:
@mundaym do you need this in branch 2.x? assuming it cherry picks I think it's reasonable to back port

mundaym commented:
@srowen I don't have any particular need to backport this. Happy to backport it if you think it is worthwhile though (not sure what the process is for that).

srowen commented:
Merged to master/3.0
